### PR TITLE
[AIRFLOW-4741] Include Sentry into core Airflow

### PR DESCRIPTION
**Pull request by tiopi**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title:

https://issues.apache.org/jira/browse/AIRFLOW-4741

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
This PR is to add Sentry SDK functionality to Airflow. The functionality includes adding tags per task event and breadcrumbs of previous tasks that ran on the DAG. 

Differences between #5416 and #5407:
#5416:  PR focuses on adding Sentry functionality to specific DAGs. This is achieved by adding a sentry hook to the contrib folder. Users can then simply call the hook with their `SENTRY_DSN` for each DAG.
#5407: PR focuses on adding Sentry to the Airflow ecosystem. Its aim is to handle all errors within airflow (Scheduler, worker, DAGs, tasks).  (This is the continuation of #5382, which I closed by accident.)

### Tests

- [x] My PR adds unit tests for `sentry.py`.

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation under `error.rst`.

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by pgagnon** [comment]
@tiopi CI is sad due to a couple of pylint errors: https://travis-ci.org/apache/airflow/jobs/544497697

edit: Wrong mentions. Sorry! My github UI bugged out. 😥

**Comment by tiopi** [comment]
> @tiopi CI is sad due to a couple of pylint errors: https://travis-ci.org/apache/airflow/jobs/544497697
> 
> edit: Wrong mentions. Sorry! My github UI bugged out. 😥

I got the tests to finally pass!

File: docs/errors.rst
Comment (pgagnon) [inline docs/errors.rst]: I think it might be preferable to add this as a parameter in `airflow.cfg` instead of relying on environment variables for configuring airflow. I know this might be a bit of a refactor but I think it would provide a much more consistent experience for the users.

File: scripts/ci/pylint_todo.txt
Comment (pgagnon) [inline scripts/ci/pylint_todo.txt]: I think it would be preferable to fix the pylint warnings instead of adding the new files to the todo list. Please take this comment with a grain of salt though, I have been guilty of the same thing. 👼

**Review by pgagnon — COMMENTED** [review]
Cursory glance.

File: docs/errors.rst
Comment (ashb) [reply docs/errors.rst]: Or if it is very common to set SENTRY_DSN then supporting both from the config and from _this_ environment variable might be the way to go.

Comment (pgagnon) [reply docs/errors.rst]: Yes; that sounds best actually. 👍

From the sentry docs [\[1\]](https://docs.sentry.io/error-reporting/configuration/?platform=python#dsn):
> If this value is not provided, the SDK will try to read it from the SENTRY_DSN environment variable.

Comment (tiopi) [reply docs/errors.rst]: I fixed it so it checks the configuration file first and then defaults to the environment variable. I also updated the documentation to reflect this.

File: airflow/config_templates/default_airflow.cfg
Comment (pgagnon) [inline airflow/config_templates/default_airflow.cfg]: Are there more configuration parameters related to sentry that you might see being added in the future? If so it might be appropriate to move this setting to a new `[sentry]` section in the config.

If that is the case it would be preferable to do this now before this is merged so that backward-compatibility concerns are minimized.

File: airflow/sentry.py
Comment (pgagnon) [inline airflow/sentry.py]: I think it may be better to use `log.debug` instead for this message.

File: airflow/config_templates/default_airflow.cfg
Comment (tiopi) [reply airflow/config_templates/default_airflow.cfg]: I really like that idea.

File: airflow/sentry.py
Comment (mik-laj) [inline airflow/sentry.py]:
```suggestion
log = logging.getLogger(__name__)
```
I think, you should use LoggingMixin.

Comment (tiopi) [reply airflow/sentry.py]: Done

File: airflow/sentry.py
Comment (pgagnon) [inline airflow/sentry.py]: Is there a reason why `CeleryIntegration` is mutually exclusive with `LoggingIntegration` in Sentry?

File: airflow/sentry.py
Comment (pgagnon) [inline airflow/sentry.py:102]: I am not familiar with sentry internals, but how will exceptions generated by task instances be contained in the sentry scope that you are opening here?

File: airflow/sentry.py
Comment (pgagnon) [inline airflow/sentry.py]: Is this function necessary? It looks like you are only using it in `add_breadcrumbs` which is in turn only used in `TaskInstance._run_raw_task`, where the TI object is already available. Passing that object would save a round-trip to the database.

Comment (tiopi) [reply airflow/sentry.py]: For this function, I try to get the previous TaskInstances. `TaskInstance._run_raw_task` only gives me the current TaskInstance running.

Comment (pgagnon) [reply airflow/sentry.py]: Ah, I see what you are doing with it.

I have a couple of loosely related comments:

- At a very high level, what information are you attempting to report to sentry with this breadcrumb? At first glance it seems that this is attempting to report the status of this task instance's dependencies, but I am not sure that this is particularly informative as a TI will not be run if its dependencies are unmet.
- How are these breadcrumbs reconciled with specific task instances? I can't seem to find any kind of scope or context linking them together. (this is basically the same question as https://github.com/apache/airflow/pull/5407#discussion_r297693203)
- Do you think it could be helpful to pass more TI data to sentry by using the breadcrumbs `data` field?
- I could see some value in reporting data about *all* TIs in a DagRun (not just this specific TI's direct parents), for instance this could help users investigate race conditions between parallel tasks. WDYT?
- I think you may be able to refactor `get_task_instance` into `get_task_instances`; if you make the function accept a list of tasks and return a list of TIs you could move it outside the loop in `sentry.py:132` which would reduce the number of calls to the database.

File: airflow/sentry.py
Comment (tiopi) [reply airflow/sentry.py]: I fixed this, since `LoggingIntegration` is set by default. The only change would be setting the level, however this is unnecessary.

File: airflow/sentry.py
Comment (tiopi) [reply airflow/sentry.py:102]: You make a good point. I had it such that the tags would overwrite each other, in the Celery case since each execution was run in a different process, tags would not leak. However, for other executors, tags would leak into other exceptions.

File: airflow/sentry.py
Comment (tiopi) [reply airflow/sentry.py]: * The breadcrumb initially was meant to recreate the dag for the specific task. However, I see more value in adding more information and adding more tasks to the breadcrumbs.
* I was doing `message` field for the reason that if there were lots of upstream tasks, the messages would not take as much space as the data table does. However, I do see more value in `data`.
* I was thinking about passing all tasks that have completed (Success or Failure) into the breadcrumbs, but ended up going with only upstream tasks as a starting point since they are the ones that affect the current task the most. However, I changed it to include all tasks that have completed.
* Refactoring of `get_task_instance` to `get_task_instances` is done!

File: scripts/ci/pylint_todo.txt
Comment (tiopi) [reply scripts/ci/pylint_todo.txt]: Fixed the Pylint errors

File: airflow/sentry.py
Comment (untitaker) [inline airflow/sentry.py:148]: Don't provide `e` here (bare call without arguments) and use a bare `raise` to preserve stacktraces in Python 2

Comment (mik-laj) [reply airflow/sentry.py:148]: We dropped support for Python. 2 on master.

**Comment by houqp** [comment]
Is this PR still blocked on any requested change? If so, I would like to help so it can be merged for our internal use.

**Comment by ashb** [comment]
It's blocked on me finding the time to review it (but a lot of my available time has been taking up with Release Management work).

File: setup.py
Comment (ashb) [inline setup.py:256]: I don't see blinker used anywhere?

Comment (untitaker) [reply setup.py:256]: Flask has a signals system that the SDK uses to hook into Flask apps. The signalling system requires blinker. It's an optional dependency of Flask.

File: airflow/sentry.py
Comment (jmcarp) [inline airflow/sentry.py]: Do we need to store these attributes on the instance? Maybe I'm missing something, but it seems like it would be simpler to pass `task_instance` to both `add_tagging` and `add_breadcrumbs` and not keep state on the instance at all.

File: airflow/sentry.py
Comment (jmcarp) [inline airflow/sentry.py]: No need to make `ti` a variable--we can just return the result of this statement.

File: airflow/sentry.py
Comment (jmcarp) [inline airflow/sentry.py]: Why would `ti` be `None`? Also, if we only care about task instances with the `SUCCESS` or `FAILED` state, can we add that filter to `get_task_instances`?

Comment (jmcarp) [reply airflow/sentry.py]: Also, the code is calling these upstream tasks, but it looks like the filter checks for all successful or failed tasks in the dag run. I think that could include tasks that aren't upstream of the relevant task instance. It might be more accurate to use something like `get_flat_relative_ids` instead.

Comment (pgagnon) [reply airflow/sentry.py]: @jmcarp 

> Also, the code is calling these upstream tasks, but it looks like the filter checks for all successful or failed tasks in the dag run. I think that could include tasks that aren't upstream of the relevant task instance. It might be more accurate to use something like `get_flat_relative_ids` instead.

This is what this PR did before but we changed it to include all tasks in the dag run, as this could give context if an error occurred due to a race condition with a parallel task,

Comment (jmcarp) [reply airflow/sentry.py]: OK, then maybe we should keep the behavior but use a clearer name--how about `completed_tasks`?

File: airflow/sentry.py
Comment (tiopi) [reply airflow/sentry.py]: Sounds good

File: airflow/sentry.py
Comment (tiopi) [reply airflow/sentry.py]: Done

File: airflow/sentry.py
Comment (tiopi) [reply airflow/sentry.py]: I changed the name to reflect that they are completed tasks.

**Comment by houqp** [comment]
Looks like all review feedbacks have been addressed, any blocker I could help resolve to get it merged?

**Comment by tiopi** [comment]
cc @jmcarp @ashb

File: docs/errors.rst
Comment (pgagnon) [inline docs/errors.rst:32]: I think you could rephrase this. It's not immediately clear to me reading this how to actually configure this integration.

I would suggest explicitely naming the actual configuration properties to be filled (`sentry_dsn`) in `airflow.cfg`, and adding an example DSN with dummy values.

**Review by pgagnon — COMMENTED** [review]
I have done a couple of passes on this and it's LGTM to me at this point barring a minor doc nit.

@ashb WDYT?

File: airflow/sentry.py
Comment (ashb) [inline airflow/sentry.py]: How is this different to the already existing `dag.get_task_instances` method?

File: airflow/sentry.py
Comment (ashb) [inline airflow/sentry.py:126]: This function isn't clear to me what it's doing (I'm possibly not that used to Sentry).

Why does it take a single task_instance, but then walk over all the task instances of the dag?

Comment (ashb) [reply airflow/sentry.py:126]: It's also not clear to me why we want all the completed TIs for a task. 

Can you show where this shows up in Sentry?

Do we instead here want to look at just the upstrream tasks of this task? Do we actually need more than the DAG_id in the breadcrumb?

File: airflow/sentry.py
Comment (ashb) [inline airflow/sentry.py]: Format how/why? This doc comment doesn't add anything to the function name.

File: airflow/sentry.py
Comment (ashb) [inline airflow/sentry.py]:
```suggestion
                    return func(task_instance, *args, session=session, **kwargs)
```

File: airflow/sentry.py
Comment (ashb) [inline airflow/sentry.py]:
```suggestion
                    raise
```

Otherwise the stack trace for the exception will be altered.

File: airflow/sentry.py
Comment (ashb) [inline airflow/sentry.py]: Why Tornado? Celery doens't use Tornado does it?

File: docs/errors.rst
Comment (ashb) [inline docs/errors.rst]: Again, not familiar with Sentry, but the dag_id and execution date will be the same as for the task that failed, so why do we need these again?

(Maybe "breadcrumbs" aren't what I thought they would be from other uses of them)

A link to the right section of the sentry docs might help here.

**Review by ashb — CHANGES_REQUESTED** [review]
Sorry it took me so long to review this.

 I have some questions about the data we sent to Sentry that I think would be cleared up by some example screenshots.

File: airflow/sentry.py
Comment (pgagnon) [reply airflow/sentry.py:126]: > It's also not clear to me why we want all the completed TIs for a task.

IIRC I suggested this. Rationale was that this could help diagnose race condition issues between related tasks.

File: airflow/sentry.py
Comment (tiopi) [reply airflow/sentry.py]: I had it because tornado is installed alongside Celery, https://github.com/apache/airflow/blob/master/setup.py#L166. However, @nathan-heskia removed the `TornadoIntegration()` since it is not essential for Celery.

File: airflow/sentry.py
Comment (tiopi) [reply airflow/sentry.py]: `_run_raw_task` does not return anything from this call. Would the return be helpful for future compatibilities?

Comment (ashb) [reply airflow/sentry.py]: Yeah - since this is a generic wrapper it's safer to return here.

File: airflow/sentry.py
Comment (ashb) [reply airflow/sentry.py:126]: Okay, that makes some sense. I feel a screen shot or two in the doc might help this.

File: docs/errors.rst
Comment (tiopi) [reply docs/errors.rst]: The dag_id, execution date should not be necessary here since they match with the respective tags.
The execution date should be replaced by a tag that captures the exact time the operation happened. This is to know if a delay on a previous operator might be causing the current one to fail.

Docs for breadcrumbs: https://docs.sentry.io/enriching-error-data/breadcrumbs/?platform=python.

File: airflow/sentry.py
Comment (tiopi) [reply airflow/sentry.py]: ~I will change it with `dagrun.get_task_instances` since `dag.get_task_instances` only allows me to capture one type of state.~ I changed `dag.get_task_instances` to accept multiple states.

File: airflow/sentry.py
Comment (tiopi) [reply airflow/sentry.py:126]: Breadcrumbs
Here is an example of the breadcrumbs in action. They give extra details on the duration of previous and concurrent task instances.

Comment (ashb) [reply airflow/sentry.py:126]: @tiopi I can't see that image (anymore?)

**Comment by houqp** [comment]
we are almost there! @tiopi let me know if there is anything I can help to get it over the finish line.

File: airflow/models/taskinstance.py
Comment (ashb) [inline airflow/models/taskinstance.py]: I think I'd like to come up with a better name for this wrapper than `format_run_task` (it's the `format` part I don't like as I don't know what this _does_)

So what does this decorator do?

Is `@Sentry.record_tags` or `@Sentry.track_context` a better name?

File: airflow/sentry.py
Comment (ashb) [inline airflow/sentry.py]: If you make 

```python
class ConfiguredSentry(DummySentry):
```

then I think this could become

```python
Sentry = DummySentry  # type: DummySentry
```

File: docs/errors.rst
Comment (ashb) [inline docs/errors.rst]:
```suggestion
.. code-block:: ini
```

File: docs/errors.rst
Comment (ashb) [inline docs/errors.rst:56]: Is it worth linking to https://docs.sentry.io/enriching-error-data/breadcrumbs/?platform=python here?

```suggestion
When a task fails with an error `breadcrumbs <https://docs.sentry.io/enriching-error-data/breadcrumbs/?platform=python>`__ will be added for the other tasks in the current dag run.

```

File: docs/errors.rst
Comment (ashb) [inline docs/errors.rst]: The code I think adds these under `completed_tasks` not `upstream_tasks`

File: airflow/sentry.py
Comment (ashb) [inline airflow/sentry.py]: Is it worth adding in `try_number` too?

**Review by ashb — CHANGES_REQUESTED** [review]

File: airflow/models/taskinstance.py
Comment (ashb) [reply airflow/models/taskinstance.py]: Or `@Sentry.enrich_errors`?

File: airflow/sentry.py
Comment (tiopi) [reply airflow/sentry.py]: Done!

File: docs/errors.rst
Comment (tiopi) [reply docs/errors.rst]: I forgot to change these

File: airflow/models/taskinstance.py
Comment (tiopi) [reply airflow/models/taskinstance.py]: `@Sentry.enrich_errors` sounds really good!

**Review by ashb — APPROVED** [review]
It took many months (sorry!) but we got there in the end!

**Comment by houqp** [comment]
great job @tiopi!

**Pull request merged**
